chromadb/db/mixins/embeddings_queue.py

```diff
@@ -243,6 +291,28 @@ def unsubscribe(self, subscription_id: UUID) -> None:
                         del self._subscriptions[topic_name]
                     return
 
+    @trace_method("SqlEmbeddingsQueue.ack", OpenTelemetryGranularity.ALL)
+    @override
+    def ack(self, subscription_id: UUID, up_to_seq_id: SeqId) -> None:
```

Comment:
This feels weird as we re inverting the responsibility of who maintains the max_seq_id table. In fact this duplicates what the metadata segment already does. Doesn't it make sense to let segments maintain their own max_seq_ids and have just an in-memory representation held by the embedding queue?

Reply:
Segments already maintain their `max_seq_id`:

- Metadata - https://github.com/chroma-core/chroma/blob/28b37392594dd7ba60e35c53f098d7f88a9d3988/chromadb/segment/impl/metadata/sqlite.py#L483-L495
- HNSW with the pickled metadata - https://github.com/chroma-core/chroma/blob/28b37392594dd7ba60e35c53f098d7f88a9d3988/chromadb/segment/impl/vector/local_persistent_hnsw.py#L213-L222

My concern was whether it is the embedding queue responsibility to update `max_seq_id` table or leave that to each segment. Instead, let segments report their max sequence IDs either via the `ack()` or simply the return of the `notify_one` and keep these in-memory (aka as attributes for each subscriber).

Reply:
Right, I see. There's two main issues with keeping max sequence IDs in memory on the embeddings queue:

- It assumes all segments have subscribed to the embeddings queue before `clean_log()` is called. This may be true in the vast majority of cases but it feels like that's a correctness bug waiting to happen. You could add an assertion that `len(subscribers) == 2` but imo that's not materially better than just persisting the max sequence ID.
- It requires segments to be loaded into memory for `clean_log()` to work. In most cases the segments will probably already be in-memory, but this could make `chroma vacuum` slower and implementation a little annoying.

Reply:
You don't have to keep max_seq_id in memory, but only query it and let segments update `max_seq_id` table on their own (aka responsibility for persisting the max_seq_id lies within the owner of the counter - each segment).

Here's a hypothetical situation situation:

- Add a large batch (that overflows the threshold)
- Vector segment updates its metadata to store the max_seq_id
- Embedding queue fails to update the `max_seq_id` - not very likely, but also not impossible. We roll back the whole shenanigan, but you still have vectors added to the index. This, too, is a correctness bug waiting to happen :)

I hope you see my point about the responsibility of segments to manage their own `max_seq_id` and make the embedding queue only a consumer (aka only query) the state upon `clean_log()`.

Reply:
> Embedding queue fails to update the max_seq_id - not very likely, but also not impossible. We roll back the whole shenanigan, but you still have vectors added to the index. This, too, is a correctness bug waiting to happen :)

Sorry, not sure how I see this could result in a correctness bug--at worst the embeddings queue won't prune processed WAL records?

> I hope you see my point about the responsibility of segments to manage their own max_seq_id and make the embedding queue only a consumer (aka only query) the state upon clean_log().

To clarify, in this implementation would `clean_log()` inspect the `max_seq_id` and/or the pickled metadata file?

Reply:
After thinking through this some more, I'd be fine having segments be responsible for updating the `max_seq_id` table themselves. We then don't need the `ack()` method. The only thing that's a little weird then is that `clean_log()`will be directly accessing the `max_seq_id` table.

Maybe we have a new component, `LogPosition`, that owns reads/writes for `max_seq_id` and we put `clean_log()` there? That seems fairly clean.

Reply:
> ...segments be responsible for updating the max_seq_id table themselves. We then don't need the ack() method. The only thing that's a little weird then is that clean_log() will be directly accessing the max_seq_id table.

I *think* this is a slight improvement. Comparison in 41a71316. End of the comment has some of the considerations I've had; I think this approach does probably balance them slightly better.

> Maybe we have a new component, LogPosition, that owns reads/writes for max_seq_id and we put clean_log() there? That seems fairly clean.

Spent a few minutes prototyping this:

1. Added a new `SqliteDB` mixin, `SqlSegmentLogPosition`.
2. The mixin requires `clean_log()`and `ack()` to be defined on the abstract `SqlDB` to avoid a circular import (👎)

--------------------------------

For the sake of writing things down, some of the things I've been considering throughout the iterations of this implementation:

- `clean_log()` assumes that `max_seq_id` entries are kept up-to-date (suggests that mutation logic for `max_seq_id` should be co-located with `clean_log()`
- `clean_log()` and `max_seq_id` are only applicable for single node (suggests that Consumer/Producer interfaces should not be modified)
- "subscribers" are ephemeral, while "segments" are persistent--`max_seq_id` has one row per segment, not one row per subscriber
- we should be able to get the `max_seq_id` of segments without having to load them into memory

I'd really like to just commit to a path at the point and stop bike shedding. It's not solely the fault of this thread, this feature has just been very delayed after chasing down the various bugs that fell out and the multiple permutations of the CIP (although I do think the user-facing side of this is far better after going through those permutations!).